ambari-server/src/main/java/org/apache/ambari/server/controller/utilities/NodeRefresher.java

```diff
@@ -0,0 +1,147 @@
+import static org.apache.ambari.server.Role.NAMENODE;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.ambari.server.AmbariException;
+import org.apache.ambari.server.actionmanager.ActionManager;
+import org.apache.ambari.server.actionmanager.RequestFactory;
+import org.apache.ambari.server.actionmanager.Stage;
+import org.apache.ambari.server.actionmanager.StageFactory;
+import org.apache.ambari.server.controller.ActionExecutionContext;
+import org.apache.ambari.server.controller.AmbariCustomCommandExecutionHelper;
+import org.apache.ambari.server.controller.internal.RequestResourceFilter;
+import org.apache.ambari.server.controller.internal.RequestStageContainer;
+import org.apache.ambari.server.events.HostsRemovedEvent;
+import org.apache.ambari.server.events.publishers.AmbariEventPublisher;
+import org.apache.ambari.server.state.Cluster;
+import org.apache.ambari.server.state.ServiceComponent;
+import org.apache.ambari.server.utils.StageUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Sends REFRESH_NODE command to the NAMENODE after deleting a host that was running a DATANODE
+ */
+@Singleton
+public class NodeRefresher {
+  private final static Logger LOG = LoggerFactory.getLogger(NodeRefresher.class);
+  private static final String REFRESH_NODE = "REFRESH_NODES";
+  private static final String REQUEST_CONTEXT = "Refresh NameNode";
+  private final AmbariCustomCommandExecutionHelper executionHelper;
+  private final StageFactory stageFactory;
+  private final ActionManager actionManager;
+  private final RequestFactory requestFactory;
+
+  @Inject
+  public NodeRefresher(AmbariCustomCommandExecutionHelper executionHelper, StageFactory stageFactory, ActionManager actionManager, RequestFactory requestFactory) {
+    this.executionHelper = executionHelper;
+    this.stageFactory = stageFactory;
+    this.actionManager = actionManager;
+    this.requestFactory = requestFactory;
+  }
+
+  public void register(AmbariEventPublisher eventPublisher) {
+    eventPublisher.register(this);
+  }
+
+  @Subscribe
+  public void onHostRemoved(HostsRemovedEvent event) {
+    if (event.hasComponent(DATANODE)) {
+      for (Cluster cluster : event.getClusters()) {
+        try {
+          LOG.info("Sending {} command to NAMENODE after host was removed {}", REFRESH_NODE, event);
+          refresh(nameNode(cluster), cluster);
+        } catch (AmbariException e) {
+          LOG.warn("Could not send " + REFRESH_NODE + " to NAMENODE after deleting host: " + event, e);
+        }
+      }
+    }
+  }
+
+  private ServiceComponent nameNode(Cluster cluster) throws AmbariException {
+    return cluster.getService("HDFS").getServiceComponent(NAMENODE.name());
+  }
+
+  protected void refresh(ServiceComponent namenode, Cluster cluster) throws AmbariException {
```

Comment:
We've been operating without this option for years without a problem.  I think there should be a configuration knob to control this behavior since NN is going to eventually reconcile dead DN anyway.

Reply:
@ncole, can you add a comment to the bugdb jira? I'm happy to not add code if it's not absolutely needed.